### ISSUE TITLE
S3 download failure crashes app

### ISSUE THREAD

**Issue by JamesMcIntosh** [aws-amplify/amplify-flutter#140]
**Describe the bug**
When a download operation is failing the App is crashing because the a message reply has already been sent.

**To Reproduce**

```
final String path = Directory.systemTemp.createTempSync().path + "/tmpfile";
final DownloadFileOptions options = S3DownloadFileOptions(accessLevel: StorageAccessLevel.protected, targetIdentityId: "ABC123");
    final DownloadFileResult result = await Amplify.Storage.downloadFile(key: "cannot_access.key", local: File(path), options: options);
```

```
Shutting down VM
FATAL EXCEPTION: main
Process: com.myovolt.myovolt, PID: 22828
java.lang.IllegalStateException: Reply already submitted
	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:139)
	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.error(MethodChannel.java:240)
	at com.amazonaws.amplify.amplify_storage_s3.AmplifyStorageOperations$StorageOperations$prepareError$1.run(AmplifyStorageOperations.kt:208)
	at android.os.Handler.handleCallback(Handler.java:914)
	...
```

To see this in action put attach the Android debugger with a breakpoint in error closure for
[downloadFile()](https://github.com/aws-amplify/amplify-flutter/blob/5a5cef48784884c6fc57d48675af69c0566e9cab/packages/amplify_storage_s3/android/src/main/kotlin/com/amazonaws/amplify/amplify_storage_s3/AmplifyStorageOperations.kt#L136) in `AmplifyStorageOperations.kt`

The first time prepareError is called it is with the S3 not found exception.
```
accept:136, AmplifyStorageOperations$StorageOperations$downloadFile$2 (com.amazonaws.amplify.amplify_storage_s3)
accept:30, AmplifyStorageOperations$StorageOperations$downloadFile$2 (com.amazonaws.amplify.amplify_storage_s3)
onError:195, AWSS3StorageDownloadFileOperation$DownloadTransferListener (com.amplifyframework.storage.s3.operation)
run:288, TransferStatusUpdater$4 (com.amazonaws.mobileconnectors.s3.transferutility)
handleCallback:914, Handler (android.os)
```

Second time it is called because of the "Failed" state change listened to by the 
[AWSS3StorageDownloadFileOperation$DownloadTransferListener](https://github.com/aws-amplify/amplify-android/blob/b0a71c80a2a0c416fa1f28f22e6f43d7114b7de7/aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StorageDownloadFileOperation.java#L166)
```
accept:136, AmplifyStorageOperations$StorageOperations$downloadFile$2 (com.amazonaws.amplify.amplify_storage_s3)
accept:30, AmplifyStorageOperations$StorageOperations$downloadFile$2 (com.amazonaws.amplify.amplify_storage_s3)
onStateChanged:176, AWSS3StorageDownloadFileOperation$DownloadTransferListener (com.amplifyframework.storage.s3.operation)
run:206, TransferStatusUpdater$2 (com.amazonaws.mobileconnectors.s3.transferutility)
handleCallback:914, Handler (android.os)
```

**Expected behavior**
Not to crash the app

**Fix**
It looks like expected behaviour of the `AWSS3StorageDownloadFileOperation` to notify the second failure is part of the Android implementation the easiest fix would be to check if the `flutterResult` has already been completed before calling `flutterResult.error()`in `com.amazonaws.amplify.amplify_storage_s3.AmplifyStorageOperations$StorageOperations$prepareError`

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[/] Android
[?] iOS - haven't tested

**Comment by JamesMcIntosh** [comment]
@Ashish-Nanda Any chance of getting this fixed as it is still a problem and the solution is pretty simple?

**Comment by JamesMcIntosh** [comment]
@Ashish-Nanda Using a var to store the state of the response such as this does work to solve the problem, I have taken to just logging the error message for failed result/error messages. Is this solution suitable?
```
fun downloadFile(@NonNull flutterResult: MethodChannel.Result, @NonNull request: Map<String, *>) {
    if (FlutterDownloadFileRequest.isValid(request)) {
        val req = FlutterDownloadFileRequest(request)
        try {
            var responded = false
            Amplify.Storage.downloadFile(req.key,
                    req.file,
                    req.options,
                    { result ->
                        if (!responded) {
                            responded = true
                            prepareDownloadFileResponse(flutterResult, result)
                        } else {
                            LOG.error("Undeliverable success, reply has already been delivered")
                        }
                    },
                    { error ->
                        if (!responded) {
                            responded = true
                            prepareError(flutterResult, FlutterStorageErrorMessage.DOWNLOAD_FILE_OPERATION_FAILED.name, error.localizedMessage, error.recoverySuggestion)
                        } else {
                            LOG.error("Undeliverable error", error)
                        }
                    }
            )
        } catch (e: Exception) {
            prepareError(flutterResult, FlutterStorageErrorMessage.DOWNLOAD_FILE_OPERATION_FAILED.name, e.localizedMessage, "")
        }
    } else {
        prepareError(flutterResult, FlutterStorageErrorMessage.DOWNLOAD_FILE_REQUEST_MALFORMED.name, "The request received was malformed", "Please ensure the request matches the method signature")
    }

}
```

**Comment by JamesMcIntosh** [comment]
@Ashish-Nanda it's been 5 months, are you happy with the solution I have suggested?

**Comment by Ashish-Nanda** [comment]
@JamesMcIntosh 
Thanks for your patience, I think this issue slipped under the radar as we've been prioritizing issues and feature requests based on activity, upvotes, etc.

I think your suggestion is in line with what we do in the case of uploadFile as well: https://github.com/aws-amplify/amplify-flutter/blob/master/packages/amplify_storage_s3/android/src/main/kotlin/com/amazonaws/amplify/amplify_storage_s3/AmplifyStorageOperations.kt#L50

Would you like to submit a PR for the fix similar to the approach above? If not I can prioritize this for our upcoming release. Thanks again for following up on this.

**Comment by JamesMcIntosh** [comment]
Thanks @Ashish-Nanda, I've created a PR #456

**Comment by haverchuck** [comment]
@JamesMcIntosh Your PR was included in the 0.1.1 release.  Thank you for contributing!

**Comment by Ashish-Nanda** [comment]
@JamesMcIntosh Please let us know if there are still any issues and we can re-open this if needed.
